File: 2023/08.17/정호조/cubeditor.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Cubeditor_1701 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String str = br.readLine();
        int ans = 0;
        for(int i=0; i<str.length(); i++){
            ans = Math.max(ans, KMP(str.substring(i)));
        }

        System.out.println(ans);
    }

    static int KMP(String pattern) {
        int[] table  = new int[pattern.length()];
        int idx =0, max = 0;

        for(int i=1; i<pattern.length(); i++){
            while (idx > 0 && pattern.charAt(i) != pattern.charAt(idx)){
                idx = table[idx - 1];
            }
            if(pattern.charAt(i) == pattern.charAt(idx)){
                idx++;
                table[i] = idx;
                max = Math.max(max, table[i]);
            }
        }

        return max;
    }
}

File: 2023/08.17/조수연/Cubeditor .kt
import java.io.*
import kotlin.math.max

lateinit var table: IntArray

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    val str = readLine()
    val n = str.length
    var answer = 0

    for (i in 0 until n) {
        val pattern = str.substring(i, n)
        answer = max(answer, makeTable(pattern))
    }

    bw.write("$answer")

    bw.flush()
    bw.close()
}

fun makeTable(pattern: String): Int {
    var idx = 0
    var answer = 0
    val len = pattern.length
    table = IntArray(len)

    for (i in 1 until len) {
        while (idx > 0 && pattern[i] != pattern[idx]) {
            idx = table[idx - 1]
        }

        if (pattern[i] == pattern[idx]) {
            table[i] = ++idx
            answer = max(answer, table[i])
        }
    }
    return answer
}